File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {

public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (!subRoot) {
            return true;
        }
        if (!root) {
            return false;
        }
        if (sameTree(root, subRoot)) { 
            return true;
        }
        else {
            return (isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot));
        }
    }
private: 
    bool sameTree(TreeNode* s, TreeNode* t) {
        if (!s && !t) {
            return true;
        }
        if (s && t && (s->val == t->val)) {
            return (sameTree(s->left, t->left) && sameTree(s->right, t->right));
        }
        return false;
    }
};

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def inOrder(self, node: TreeNode, result: list[str]):

        if not node:
            return 
        self.inOrder(node.left, result)
        result.append(node.val)
        self.inOrder(node.right, result)

    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        result = []
        self.inOrder(root, result)
        return result[k-1]

# Runtime: O(n) all nodes are processed once
# Space: O(n) array data structure is created with n elements, the same
#. as the number of nodes 

        